Leetcode/aoc_2015_d4.py

```diff
@@ -0,0 +1,140 @@
+	h = file.read()
+
+
+def find_hash(h: str) -> int:
+	# set target hash to 0 for now. placeholder so that the
+	# while loop can run once. Could compensate by doing a 
+	# do while instead of this.
+	target_hash = "0"
+	# the iterable number that will get added to the string.
+	# starts at 0 but gets added right away in the while loop
+	# so that it truly starts at 1.
+	find_hash = 0
+	# slice off just the first five characters to check if it's
+	# leading with 5 zeroes.
+	while target_hash[:5] != "00000":
+		# Starts at 1 since the answer can't start with a leading
+		# zero. And by adding the number first, we then perform the
+		# calculations, so that once the next loop starts, if the
+		# previous hash started with 5 zeroes, the next loop wouldn't
+		# run, thereby not adding one to the find_hash.
+		find_hash += 1
+		# concate the secret key and the iterable then encodes it
+		# to be turned into an md5 hash.
+		str2hash = (h + str(find_hash)).encode()
+		# gets the md5 hash, then converts it into a readable string.
+		target_hash = hashlib.md5(str2hash).hexdigest()
+
+	return find_hash
+
+'''
+ ----- Thoughts on part 1 after completing it. -----
+This one was a bit tricky. I understood it conceptually, but I didn't know
+how to put it in code as easily since I didn't know how the hashlib and md5
+worked. It took a bit of trial and error, but understanding it conceptually
+made coding it a bit easier, I think.
+Since it must go through numbers by iteration, it can take a good amount of time
+find the right string that leads with 5 zeroes.
+If I really wanted to improvet this, I could probably do something similar to my
+java mega millions program and have it multithreaded with each loop going through
+say 999_999 numbers. e.x. thread 1 does 1 to 100_000, thread 2 does 1_000_000 to
+1_999_999 numbers. But that's micro-optimization, since there's no guarantee it'd
+be a number that high. For example, my secret key had an answer that was quite low.
+'''
+
+'''
+ ----- Part 2 -----
+Part 2 is the same thing, just finding one that starts with six zeroes
+instead of five. Seeing this, I can now see that my thought on multi-threading
+would be more helpful as finding one with six leading zeroes would take much
+more time.
+Part 2 would be the same code as part 1, just changing the while conditional.
+So to take it a step further, I'll just add another parameter to the function.
+and set the while loop to use that.
+
+Knowing that this will take longer, I have decided to go back in both
+part 2 and part 1 to get execution time.
+This is done by getting cpu time using the time library and using process_time()
+'''
+
+# This function can completely replace the part 1 version.
+# Works the exact same way as the first one, default is 5 leading zeroes.
+# can just do find_hash_p2(h) and it'll look for 5 leading,
+# but you can pass it a different amount of leading zeroes to look for
+# and it will look for that instead.
```

Comment:
:nise: use of a default param, good way to prevent breaking the existing API.

Reply:
:nise: